### PR TITLE
Integration ModelAPI to OV Classification task

### PR DESCRIPTION
**Pull request by akorobeinikov**
Can be merged after: [openvinotoolkit/open_model_zoo#2900](https://github.com/openvinotoolkit/open_model_zoo/pull/2900) and [openvinotoolkit/training_extensions#727](https://github.com/openvinotoolkit/training_extensions/pull/727)

### CODE REVIEW

File: torchreid/integration/sc/model_wrappers/classification.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            'multilabel': BooleanValue(default_value=False),
+            'topk': NumericalValue(value_type=int, default_value=1),
+            'labels': ListValue(description="List of class labels"),
+            'path_to_labels': StringValue(description="Path to file with labels. Overrides the labels, if they sets via 'labels' parameter")
+        })
+
+        return parameters
+
+    @staticmethod
+    def _load_labels(labels_file):
```
Comment (sovrasov) [inline torchreid/integration/sc/model_wrappers/classification.py]: This plain text format doesn't cover the hierarchical labels case. I'm afraid that we can't support it without ote_sdk dependency. If we have one, label schema could be pickled.
BTW there is activity in ote on adding a demo script that can use tasks for inference. You can synchronize with it and probably unify the way of [serializing](https://github.com/openvinotoolkit/training_extensions/pull/737) of LabelSchemaEntity object. However, in SC environment a successor of LabelSchemaEntity is substituted everywhere, so downcasting back to LabelSchemaEntity could be required.

File: torchreid/integration/sc/model_wrappers/classification.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        layer_shape = self.outputs[layer_name].shape
+
+        if len(layer_shape) != 2 and len(layer_shape) != 4:
+            raise RuntimeError('The Classification model wrapper supports topologies only with 2D or 4D output')
+        if len(layer_shape) == 4 and (layer_shape[2] != 1 or layer_shape[3] != 1):
+            raise RuntimeError('The Classification model wrapper supports topologies only with 4D '
+                               'output which has last two dimensions of size 1')
+        if self.labels:
+            if (layer_shape[1] == len(self.labels) + 1):
+                self.labels.insert(0, 'other')
```
Comment (sovrasov) [inline torchreid/integration/sc/model_wrappers/classification.py]: What does 'other' label actually represent?

File: torchreid/integration/sc/openvino_task.py [hunk 3]
```diff
@@ -199,9 +170,40 @@ def evaluate(self,
...
         if evaluation_metric is not None:
             logger.warning(f'Requested to use {evaluation_metric} metric,'
-                            'but parameter is ignored. Use accuracy instead.')
+                           'but parameter is ignored. Use accuracy instead.')
         output_result_set.performance = MetricsHelper.compute_accuracy(output_result_set).get_performance()
 
+    def deploy(self,
+               output_path: str):
+        work_dir = os.path.dirname(demo.__file__)
+        model_file = inspect.getfile(type(self.inferencer.model))
```
Comment (sovrasov) [inline torchreid/integration/sc/openvino_task.py:173]: Seems like code in this method is mostly algo/model agnostic, so I guess it can be moved to a higher level entity.
Also I heard that Greg is working on some architecture proposals touching that stuff. May be it worth to reach him and discuss.

File: torchreid/integration/sc/parameters.py [hunk 4]
```diff
@@ -56,6 +56,32 @@ class __LearningParameters(ParameterGroup):
...
+                                editable=False,
+                                visible_in_ui=True)
+
+        @attrs
+        class __Postprocessing(ParameterGroup):
+            header = string_attribute("Postprocessing")
+            description = header
+
+            topk = configurable_integer(
+                header="Top k elements",
```
Comment (sovrasov) [inline torchreid/integration/sc/parameters.py]: That's not the SC style of visualization/not applicable for multilabel and hierarchical cases. Perhaps one could try web UI on a working SC instance to get some inspiration if the the exportable code aims to mimic it somehow.

File: torchreid/integration/sc/parameters.py [hunk 5]
```diff
@@ -56,6 +56,32 @@ class __LearningParameters(ParameterGroup):
...
+    @attrs
+    class __InferenceParameters(ParameterGroup):
+        header = string_attribute("Parameters for inference")
+        description = header
+
+        class_name = selectable(default_value=Models.CLASSIFICATION,
+                                header="Model class for inference",
+                                description="Model classes with defined pre- and postprocessing",
+                                editable=False,
+                                visible_in_ui=True)
```
Comment (sovrasov) [inline torchreid/integration/sc/parameters.py]: `class_name` looks like an internal implementation parameter, which is not configurable. Why do we need it here?

File: torchreid/integration/sc/openvino_task.py [hunk 6]
```diff
@@ -95,59 +90,34 @@ def __init__(
...
-        self.n, self.c, self.h, self.w = self.net.input_info[self.input_blob_name].tensor_desc.dims
-        self.keep_aspect_ratio_resize = False
-        self.pad_value = 0
-        self.multilabel = multilabel
         self.empty_label = empty_label
-
-    @staticmethod
-    def resize_image(image: np.ndarray, size: Tuple[int], keep_aspect_ratio: bool = False) -> np.ndarray:
-        if not keep_aspect_ratio:
-            img = PIL.Image.fromarray(image).resize(size, resample=PIL.Image.BILINEAR)
```
Comment (sovrasov) [inline torchreid/integration/sc/openvino_task.py]: Classification models are trained with PIL resize, so expect an accuracy drop if `Classification(ImageModel)` is replacing it with cv.resize.

File: torchreid/integration/sc/model_wrappers/classification.py (hunk 1 above)
Comment (akorobeinikov) [reply torchreid/integration/sc/model_wrappers/classification.py]: I am aware of `LabelSchemaEntity` and it can have any structure. But as I see any inferencer using `List[LabelEntity]` and it is meaning that we already transpose complex label schema structure to linear structure and what about `LabelEntity` it have only one important field - name. So I think current implementation for openvino inferencer is cover label cases, because before saving labels to package or creating model from ModelAPI were called function `get_labels` for `task_environment.label_schema`. Maybe I don’t see the whole picture, in this case I can add serializing and according reading to demo or model wrappers.

File: torchreid/integration/sc/openvino_task.py (hunk 3 above)
Comment (akorobeinikov) [reply torchreid/integration/sc/openvino_task.py:173]: I agree that code is mostly agnostic, but this depends on task environment as well as `export` and `optimize`. Export and optimize as much as deploy almost the same for all tasks. Interfaces for these methods contains only 1 abstract method. So at the moment OTE_SDK doesn't have any higher level entity. So now I am going create deploy for all tasks. If I  concoct new level I'll realize it. What about Greg, I add interface suggested by him in [PR](https://github.com/openvinotoolkit/training_extensions/pull/727) to ote

File: torchreid/integration/sc/parameters.py (hunk 5 above)
Comment (akorobeinikov) [reply torchreid/integration/sc/parameters.py:68]: `class_name` it is type of using model wrapper. For any new model with own pre- and postprocessing you create `template.yml` file in which you should override this parameter. So I add `class_name ` to `configuration.yaml` and `parameters.py`

File: torchreid/integration/sc/utils.py [hunk 7]
```diff
@@ -166,29 +166,6 @@ def is_multilabel(self):
         return self.multilabel
 
 
-def get_empty_label(label_schema: LabelSchemaEntity) -> LabelEntity:
```
Comment (sovrasov) [inline torchreid/integration/sc/utils.py:169]: These functions are used by OTEClassificationInferenceTask, so I guess they can't be just removed.

File: torchreid/integration/sc/inference_task.py [hunk 8]
```diff
@@ -42,14 +42,13 @@
 from ote_sdk.usecases.tasks.interfaces.export_interface import ExportType, IExportTask
 from ote_sdk.usecases.tasks.interfaces.inference_interface import IInferenceTask
 from ote_sdk.usecases.tasks.interfaces.unload_interface import IUnload
+from ote_sdk.utils.labels_utils import get_empty_label, get_leaf_labels, get_ancestors_by_prediction
 from scripts.default_config import (get_default_config, imagedata_kwargs,
                                     merge_from_files_with_base, model_kwargs)
 from torchreid.apis.export import export_ir, export_onnx
 from torchreid.integration.sc.monitors import DefaultMetricsMonitor, StopCallback
 from torchreid.integration.sc.parameters import OTEClassificationParameters
-from torchreid.integration.sc.utils import (active_score_from_probs, get_actmap,
```
Comment (sovrasov) [inline torchreid/integration/sc/inference_task.py:50]: Please revert these changes. Seems like they are just merge artifacts. Pytorch-related stuff shouldn't be affected by this PR.

File: setup.py [hunk 9]
```diff
@@ -48,7 +48,9 @@ def get_requirements(filename='requirements.txt'):
                 requires.append(line)
     return requires, links
 
+
 packages, links = get_requirements()
+packages.extend(get_requirements('modelapi-requirements.txt'))
```
Comment (sovrasov) [inline setup.py]: Is the deployment package depend on torchreid? If no, then please remove this. You could add a warning if someone tries to import Classification model wrapper without omz-modelapi installed. The same is already done for nncf.

File: torchreid/integration/sc/openvino_task.py [hunk 10]
```diff
@@ -91,72 +87,32 @@ def __init__(
...
-
-        if len(label_schema.get_labels(False)) == 1:
-            self.labels = label_schema.get_labels(include_empty=True)
-        else:
-            self.labels = label_schema.get_labels(include_empty=False)
-        self.empty_label = get_empty_label(label_schema)
-        self.multilabel = len(label_schema.get_groups(False)) > 1 and \
-                len(label_schema.get_groups(False)) == len(label_schema.get_labels(include_empty=False))
 
-        self.hierarchical = False
```
Comment (sovrasov) [inline torchreid/integration/sc/openvino_task.py:104]: I think that at the current stage of development the original OpenVINO inferencer should be restored in OpenVINO task, while omz-based inferencer could be utilized in deploy() method. 
The motivation is the following:
- these changes are breaking current functionality in SC: if someone create a project with hierarchical labels, this will cause a training pipeline failure, because after training the model is converted to IR and evaluated via OV task. So after merging this PR we would get shrinking of SC classification component capabilities. SC managers wouldn't be happy when this feature will disappear, since they love it for some reasons.
- we are going to improve hierarchical label handling by introducing multihead classifications models and more sophisticated postprocessing. OpenVINO task should be self-containing to enable experiments in that direction. At the same time basic scenarios (single/multilabel) will be backward compatible with the omz model wrapper.

**Review by sovrasov — CHANGES_REQUESTED** [review]

File: torchreid/integration/sc/utils.py (hunk 7 above)
Comment (akorobeinikov) [reply torchreid/integration/sc/utils.py:169]: I didn't remove these functions, I moved them to `ote_sdk`. Please see [PR](https://github.com/openvinotoolkit/training_extensions/pull/775)

File: torchreid/integration/sc/inference_task.py (hunk 8 above)
Comment (akorobeinikov) [reply torchreid/integration/sc/inference_task.py:50]: Above. Move to `ote_sdk`

File: torchreid/integration/sc/openvino_task.py (hunk 10 above)
Comment (akorobeinikov) [reply torchreid/integration/sc/openvino_task.py:104]: I didn't delete possibility to use hierarchical labels, I separated postprocessing for classification with any operations using labels. And I save old postprocessing in my current realization. Please take a look [PR](https://github.com/openvinotoolkit/training_extensions/pull/775)

File: setup.py (hunk 9 above)
Comment (akorobeinikov) [reply setup.py]: I add warnings, but what about SC tests. Will they not fall due to lack of modelapi dependency in `init_venv`?

**Comment by kprokofi** [comment]
@akorobeinikov , Have you finished with this PR?

**Comment by akorobeinikov** [comment]
> @akorobeinikov , Have you finished with this PR?

Yes. If you haven't objections, please merge it after merge [775](https://github.com/openvinotoolkit/training_extensions/pull/775)
